samples/matter/common/src/diagnostic/diagnostic_logs_crash.cpp

```diff
@@ -0,0 +1,215 @@
+	for (size_t i = 0; i < ARRAY_SIZE(mDescription.Esf.fpu.s); i += sizeof(uint32_t)) {
+		Verify(Collect("s[%2d]:  0x%08x  s[%2d]:  0x%08x"
+			       "  s[%2d]:  0x%08x  s[%2d]:  0x%08x\n",
+			       i, (uint32_t)mDescription.Esf.fpu.s[i], i + 1, (uint32_t)mDescription.Esf.fpu.s[i + 1],
+			       i + 2, (uint32_t)mDescription.Esf.fpu.s[i + 2], i + 3,
+			       (uint32_t)mDescription.Esf.fpu.s[i + 3]),
+		       false);
+	}
+#ifdef CONFIG_VFP_FEATURE_REGS_S64_D32
+	for (size_t i = 0; i < ARRAY_SIZE(mDescription.Esf.fpu.d); i += sizeof(uint32_t)) {
+		Verify(Collect("d[%2d]:  0x%16llx  d[%2d]:  0x%16llx"
+			       "  d[%2d]:  0x%16llx  d[%2d]:  0x%16llx\n",
+			       i, (uint64_t)mDescription.Esf.fpu.d[i], i + 1, (uint64_t)mDescription.Esf.fpu.d[i + 1],
+			       i + 2, (uint64_t)mDescription.Esf.fpu.d[i + 2], i + 3,
+			       (uint64_t)mDescription.Esf.fpu.d[i + 3]),
+		       false);
+	}
+#endif
+	Verify(Collect("fpscr:  0x%08x\n", mDescription.Esf.fpu.fpscr), false);
+#endif
+
+	return true;
+}
+
+bool CrashData::ExtraDump()
+{
+	Verify(mStage == Stage::Extra, false);
+
+#if defined(CONFIG_EXTRA_EXCEPTION_INFO)
+	const struct _callee_saved *callee = mDescription.Esf.extra_info.callee;
+
+	if (callee != NULL) {
+		Verify(Collect("r4/v1:  0x%08x  r5/v2:  0x%08x  r6/v3:  0x%08x\n", callee->v1, callee->v2, callee->v3),
+		       false);
+		Verify(Collect("r7/v4:  0x%08x  r8/v5:  0x%08x  r9/v6:  0x%08x\n", callee->v4, callee->v5, callee->v6),
+		       false);
+		Verify(Collect("r10/v7: 0x%08x  r11/v8: 0x%08x    psp:  0x%08x\n", callee->v7, callee->v8, callee->psp),
+		       false);
+	}
+
+	Verify(Collect("EXC_RETURN: 0x%0x\n", mDescription.Esf.extra_info.exc_return), false);
+
+#endif /* CONFIG_EXTRA_EXCEPTION_INFO */
+
+	return true;
+}
+
+bool CrashData::ReasonDump()
+{
+	Verify(mStage == Stage::Reason, false);
+
+	Verify(Collect("ZEPHYR FATAL ERROR %d: %s on CPU %d\n", mDescription.Reason, ReasonToStr(mDescription.Reason),
+		       _current_cpu->id),
+	       false);
+
+	return true;
+}
+
+bool CrashData::MultithreadDump()
+{
+	Verify(mStage == Stage::Multithread, false);
+
+#ifdef CONFIG_MULTITHREADING
+	if (mDescription.ThreadName && mDescription.ThreadInt) {
+		Verify(Collect("Current thread: %p (%s)\n", mDescription.ThreadInt, mDescription.ThreadName), false);
+	}
+#endif
+
+	return true;
+}
+
+const char *CrashData::ReasonToStr(unsigned int reason)
+{
+	switch (reason) {
+	case K_ERR_CPU_EXCEPTION:
+		return "CPU exception";
+	case K_ERR_SPURIOUS_IRQ:
+		return "Unhandled interrupt";
+	case K_ERR_STACK_CHK_FAIL:
+		return "Stack overflow";
+	case K_ERR_KERNEL_OOPS:
+		return "Kernel oops";
+	case K_ERR_KERNEL_PANIC:
+		return "Kernel panic";
+	default:
+		return "Unknown error";
+	}
+}
+
+bool CrashData::Collect(const char *format, ...)
+{
+	if (mLastStage > mStage) {
+		/* Last known stage is higher than current one, so skip it */
+		mLine = 0;
+		return true;
+	} else if (mLastLine >= mLine && mLastStage != Stage::None) {
```

Comment:
This condition seems to assume that `mState == mLastStage`, which is checked later. Thus, it assumes that `ProcessConversionToLog` is deterministic and always prints the same lines. If that's the case, wouldn't it be enough and easier to track `mLastLine` and not care about the current stage at all?

Reply:
that's right, I overcomplicated it.